src/main.rs

```diff
@@ -1,359 +1,117 @@
-        2 => "error",
-        3..=255 => "off",
     };
 
-    let levels = match std::env::var("LOG").ok() {
-        Some(level) => level,
-        None => match level {
-            "off" => "off".to_string(),
-            _ => [
-                format!("vector={}", level),
-                format!("codec={}", level),
-                format!("file_source={}", level),
-                format!("tower_limit=trace"),
-                format!("rdkafka={}", level),
-            ]
-            .join(",")
-            .to_string(),
-        },
-    };
-
-    let color = match opts.color.clone() {
-        #[cfg(unix)]
-        Color::Auto => atty::is(atty::Stream::Stdout),
-        #[cfg(windows)]
-        Color::Auto => false, // ANSI colors are not supported by cmd.exe
-        Color::Always => true,
-        Color::Never => false,
-    };
-
-    let json = match &opts.log_format {
-        LogFormat::Text => false,
-        LogFormat::Json => true,
-    };
-
-    trace::init(color, json, levels.as_str());
-
-    metrics::init().expect("metrics initialization failed");
-
-    sub_command.map(|s| {
-        std::process::exit(match s {
-            SubCommand::Validate(v) => validate::validate(&v, color),
-            SubCommand::List(l) => list::cmd(&l),
-            SubCommand::Test(t) => unit_test::cmd(&t),
-            SubCommand::Generate(g) => generate::cmd(&g),
-        })
-    });
+    if signal == SIGINT || signal == SIGTERM {
+        use futures01::future::Either;
 
-    info!("Log level {:?} is enabled.", level);
+        info!("Shutting down.");
+        let shutdown = topology.stop();
 
-    if let Some(threads) = opts.threads {
-        if threads < 1 {
-            error!("The `threads` argument must be greater or equal to 1.");
-            std::process::exit(exitcode::CONFIG);
+        match rt.block_on(shutdown.select2(signals.into_future())) {
+            Ok(Either::A(_)) => { /* Graceful shutdown finished */ }
+            Ok(Either::B(_)) => {
+                info!("Shutting down immediately.");
+                // Dropping the shutdown future will immediately shut the server down
+            }
+            Err(_) => unreachable!(),
         }
+    } else if signal == SIGQUIT {
+        info!("Shutting down immediately");
+        drop(topology);
+    } else {
+        unreachable!();
     }
 
-    let config_paths = config_paths::prepare(opts.config_paths.clone()).unwrap_or_else(|| {
-        std::process::exit(exitcode::CONFIG);
-    });
+    runtime.shutdown_now().wait().unwrap();
+}
 
-    if opts.watch_config {
-        // Start listening for config changes immediately.
-        vector::topology::config::watcher::config_watcher(
-            config_paths.clone(),
-            vector::topology::config::watcher::CONFIG_WATCH_DELAY,
-        )
-        .unwrap_or_else(|error| {
-            error!(message = "Unable to start config watcher.", %error);
-            std::process::exit(exitcode::CONFIG);
+#[cfg(windows)]
```

Comment:
I feel like starting as a service should be opt-in. It seems like running `vector` via Powershell after this would mean we started vector as a service, not directly? I'd prefer that do be done via the dedicated service start command if we could?

What kind of user experience difference does this bring?

Reply:
It is actually opt-in. If you start `vector` via Powershell or windows command line, it won't start as a service. However, since console mode and windows service mode have the same entry point (the `main` function), I need to check whether the binary has been started as a service since I need to register the service with the Windows Service Control Manager when it is started by Windows as a service.

Do detect whether or not `vector` has been started as a service, I then attempt to call the Win32 `StartServiceCtrlDispatcher` function (through `windows-service-rs` library) and check its result code.

This is why I first try to start as a serevice and fall-back to console mode if I fail. This is somehow counter-intuitive, I had some trouble understanding the whole principle, see for  example the [MSDN Service Sample](https://docs.microsoft.com/en-us/windows/win32/services/svc-cpp)

Reply:
OKay! This makes sense. Thanks. :)